lib/CL/devices/tbb/tbb_scheduler.cc

```diff
@@ -0,0 +1,367 @@
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+   IN THE SOFTWARE.
+*/
+
+#define _GNU_SOURCE
+
+#ifdef __linux__
+#include <sched.h>
+#endif
+
+#include <algorithm>
+#include <pthread.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <tbb/blocked_range3d.h>
+#include <tbb/parallel_for.h>
+#include <tbb/partitioner.h>
+#include <tbb/task_arena.h>
+
+#include "common.h"
+#include "pocl_cl.h"
+#include "pocl_mem_management.h"
+#include "pocl_runtime_config.h"
+#include "pocl_util.h"
+#include "tbb.h"
+#include "tbb_scheduler.h"
+#include "tbb_utils.h"
+#include "utlist.h"
+
+static void* pocl_tbb_driver_thread (void *p);
+
+typedef struct scheduler_data_
+{
+  unsigned printf_buf_size;
+  void* printf_buf_global_ptr;
+
+  size_t local_mem_size;
+  void* local_mem_global_ptr;
+
+  size_t num_tbb_threads;
+
+  _cl_command_node *work_queue __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  POCL_FAST_LOCK_T wq_lock_fast __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+
+  pthread_t meta_thread __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  pthread_cond_t wake_meta_thread __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  int meta_thread_shutdown_requested;
+} scheduler_data __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+
+static scheduler_data scheduler;
+
+/* External functions declared in tbb_scheduler.h */
+
+void
+tbb_scheduler_init (cl_device_id device)
+{
+  POCL_FAST_INIT (scheduler.wq_lock_fast);
+
+  pthread_cond_init (&(scheduler.wake_meta_thread), NULL);
+
+  scheduler.printf_buf_size = device->printf_buffer_size;
+  assert (device->printf_buffer_size > 0);
+
+  /* safety margin - aligning pointers later (in kernel arg setup)
+   * may require more local memory than actual local mem size.
+   * TODO fix this */
+  scheduler.local_mem_size = device->local_mem_size << 4;
+
+  scheduler.num_tbb_threads = device->max_compute_units;
+
+  /* alloc local memory for all threads making sure the memory for each thread is aligned. */
+  scheduler.printf_buf_size = (1 + (scheduler.printf_buf_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.local_mem_size = (1 + (scheduler.local_mem_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.printf_buf_global_ptr = pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.printf_buf_size * scheduler.num_tbb_threads);
+  scheduler.local_mem_global_ptr = pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.local_mem_size * scheduler.num_tbb_threads);
+
+  /* create one meta thread to serve as an async interface thread */
+  pthread_create (&scheduler.meta_thread, NULL, pocl_tbb_driver_thread, NULL);
+
+  /* look up each environment variable once to cache the result. */
+  if (pocl_is_option_set("POCL_TBB_GRAIN_SIZE"))
+    printf("[TBB] using a grain size of %i\n", pocl_get_int_option("POCL_TBB_GRAIN_SIZE", 1));
```

Comment:
is that how logging in pocl works?

Reply:
pocl does not log anything by default. There is POCL_DEBUG which can be set to 1 or specific categories, but I want to print this even when POCL_DEBUG is disabled because it is so important.

Reply:
would be a question for the pocl developers, also where to print it, std::err, std::out, ... I think libs should never pollute std::cout in release builds, but these things always depend on the project policies, maybe look for other outputs and check which stream they use

Reply:
Almost all other output is with `printf()` inside an `#ifdef DEBUG_MY_DRIVER` (disabled by default).
I will use the more recent `POCL_MSG_PRINT_GENERAL()` wrapper for my prints, although I feel very uncomfortable with it.